packages/react/src/index.ts

```diff
@@ -163,24 +103,138 @@ function createEffectStore() {
+//   right after.
+//
+//   The useSyncExternalStore shim will use some of these hooks when we invoke
+//   it while entering a component render. We need to prevent this dispatcher
+//   change caused by these hooks from re-triggering our entering logic (it
+//   would cause an infinite loop if we did not). We do this by using a lock to
+//   prevent the setter from running while we are in the setter.
+//
+//   When a Component's function body invokes useReducer, useState, or useMemo,
+//   this change in dispatcher should not signal that we are exiting a component
+//   render. We ignore this change by detecting these dispatchers as different
+//   from ContextOnlyDispatcher and other valid dispatchers.
+//
+// - The `use` hook will change the dispatcher to from a valid update dispatcher
+//   to a valid mount dispatcher in some cases. Similarly to useReducer
+//   mentioned above, we should not signal that we are exiting a component
+//   during this change. Because these other valid dispatchers do not pass the
+//   ContextOnlyDispatcher check, they do not affect our logic.
+let lock = false;
+let currentDispatcher: ReactDispatcher | null = null;
+Object.defineProperty(ReactInternals.ReactCurrentDispatcher, "current", {
+	get() {
+		return currentDispatcher;
+	},
+	set(nextDispatcher: ReactDispatcher) {
+		if (lock) {
+			currentDispatcher = nextDispatcher;
+			return;
 		}
 
-		if (type && typeof type === "object") {
-			if (type.$$typeof === ReactMemoType) {
-				type.type = ProxyFunctionalComponent(type.type);
-				return jsx.call(jsx, type, props, ...rest);
-			} else if (type.$$typeof === ReactForwardRefType) {
-				type.render = ProxyFunctionalComponent(type.render);
-				return jsx.call(jsx, type, props, ...rest);
+		const currentDispatcherType = getDispatcherType(currentDispatcher);
+		const nextDispatcherType = getDispatcherType(nextDispatcher);
+
+		// We are entering a component render if the current dispatcher is the
+		// ContextOnlyDispatcher and the next dispatcher is a valid dispatcher.
+		const isEnteringComponentRender =
+			currentDispatcherType === ContextOnlyDispatcherType &&
+			nextDispatcherType === ValidDispatcherType;
+
+		// We are exiting a component render if the current dispatcher is a valid
+		// dispatcher and the next dispatcher is the ContextOnlyDispatcher.
+		const isExitingComponentRender =
+			currentDispatcherType === ValidDispatcherType &&
+			nextDispatcherType === ContextOnlyDispatcherType;
+
+		// Update the current dispatcher now so the hooks inside of the
+		// useSyncExternalStore shim get the right dispatcher.
+		currentDispatcher = nextDispatcher;
+		if (isEnteringComponentRender) {
+			lock = true;
+			const storeRef = nextDispatcher.useRef<EffectStore>();
+			if (storeRef.current == null) {
+				storeRef.current = createEffectStore();
 			}
+
+			const store = storeRef.current;
+			useSyncExternalStore(
+				store.subscribe,
+				store.getSnapshot,
+				store.getSnapshot
+			);
```

Comment:
TODO: move these hook calls into a custom hook function for hopefully better grouping in devtools